### ISSUE TITLE
UnoCSS Dist-chunk mode in Astro

### ISSUE THREAD

**Issue by drumsta** [unocss/unocss#2076]
TL;DR

- UnoCSS mode `dist-chunk` sounds very promising for MPA style applications.
- I can't make it work in Astro neither in `SSG` nor in `SSR + Node adapter` mode.
- Actually, I'm not sure if there is an issue in UnoCSS, or in Astro, or even in Vite, or in combination of any of them.

I'm building MPA app with Astro + UnoCSS. Reading through available documentation, I've found `dist-chunk` mode in UnoCSS which should be a great fit for MPA style apps. Wanted to give it a try and encountered some issues.

It's just a starter app for Astro `npm create astro@latest` and UnoCSS in addition. The source code is available at [astro-unocss-sample repo](https://github.com/drumsta/astro-unocss-sample).

Using standard installation UnoCSS works just fine, `dev`, `build` & `preview` work like a charm.

Let's try to change the mode to `dist-chunk` in the configuration `astro.config.mjs`:

```
import { defineConfig } from 'astro/config';
import unocss from 'unocss/astro';

export default defineConfig({
    integrations: [
        unocss({
            mode: 'dist-chunk',
          }),
      ],
});
```

Now, running `npm run build` I get an error:

```
PS C:\Users\...\astro-unocss-sample> npm run build

> @example/minimal@0.0.1 build
> astro build

15:21:43 [build] output target: static
15:21:43 [build] Collecting build info...
15:21:43 [build] Completed in 34ms.
15:21:43 [build] Building static entrypoints...
[vite]: Rollup failed to resolve import "uno.css" from "astro:scripts/page-ssr.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
 error   [vite]: Rollup failed to resolve import "uno.css" from "astro:scripts/page-ssr.js".
  This is most likely unintended because it can break your application at runtime.
  If you do want to externalize this module explicitly add it to
  `build.rollupOptions.external`
  File:
    C:\Users\...\astro-unocss-sample\node_modules\vite\dist\node\chunks\dep-5605cfa4.js:45832:19
  Code:
    45831 |         if (!importer || !/\?commonjs-external$/.test(importer)) {
    > 45832 |             throw new Error(`[vite]: Rollup failed to resolve import "${id}" from "${importer}".\n` +
            |                   ^
      45833 |                 `This is most likely unintended because it can break your application at runtime.\n` +
      45834 |                 `If you do want to externalize this module explicitly add it to\n` +
      45835 |                 `\`build.rollupOptions.external\``);
```

Let's add `uno.css` to the `build.rollupOptions.external`:

```
import { defineConfig } from 'astro/config';
import unocss from 'unocss/astro';

export default defineConfig({
    integrations: [
        unocss({
            mode: 'dist-chunk',
        }),
    ],
    vite: {
        build: {
            rollupOptions: {
                external: ['uno.css'],
            },
        },
    },
});
```

And run `npm run build`:

```
PS C:\Users\...\astro-unocss-sample> npm run build

> @example/minimal@0.0.1 build
> astro build

16:09:54 [build] output target: static
16:09:54 [build] Collecting build info...
16:09:54 [build] Completed in 30ms.
16:09:54 [build] Building static entrypoints...
16:09:55 [build] Completed in 1.15s.

 generating static routes 
 error   Cannot find package 'uno.css' imported from C:\Users\...\astro-unocss-sample\dist\entry.mjs
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'uno.css' imported from C:\Users\...\astro-unocss-sample\dist\entry.mjs
    at new NodeError (node:internal/errors:399:5)
    at packageResolve (node:internal/modules/esm/resolve:794:9)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36)

```

By default Astro runs in `SSG` mode. Let's try to switch to Astro's `SSR` mode by installing `Node` adapter `npm install @astrojs/node`  and changing the configuration correspondingly:

```
import node from '@astrojs/node';
import { defineConfig } from 'astro/config';
import unocss from 'unocss/astro';

export default defineConfig({
    output: 'server',
    adapter: node({
        mode: 'standalone',
    }),
    integrations: [
        unocss({
            mode: 'dist-chunk',
        }),
    ],
    vite: {
          build: {
            rollupOptions: {
                external: ['uno.css'],
            },
        },
    },
});
```

And run `npm run build` again:

```
PS C:\Users\...\astro-unocss-sample> npm run build

> @example/minimal@0.0.1 build
> astro build

16:18:17 [build] output target: server
16:18:17 [build] deploy adapter: @astrojs/node
16:18:17 [build] Collecting build info...
16:18:17 [build] Completed in 34ms.
16:18:17 [build] Building server entrypoints...
16:18:20 [build] Completed in 3.36s.

 finalizing server assets 

16:18:20 [build] Rearranging server assets...
16:18:20 [build] Server built in 3.43s
16:18:20 [build] Complete!
```

Above looks promising. Let's run `npm run preview`:

```
PS C:\Users\...\astro-unocss-sample> npm run preview

> @example/minimal@0.0.1 preview
> astro preview

 error   The server entrypoint C:\Users\...\astro-unocss-sample\dist\server\entry.mjs does not exist. Have you ran a build yet?
  File:
    C:\Users\...\astro-unocss-sample\node_modules\@astrojs\node\dist\preview.js:23:13
  Code:
    22 |     if (err.code === "ERR_MODULE_NOT_FOUND") {
    > 23 |       throw new Error(
         |             ^
      24 |         `The server entrypoint ${fileURLToPath(
      25 |           serverEntrypoint
      26 |         )} does not exist. Have you ran a build yet?`

```

Let's look inside:

```
PS C:\Users\...\astro-unocss-sample\dist> ls -r

    Directory: C:\Users\...\astro-unocss-sample\dist

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          12/01/2023    16:18                client
d----          12/01/2023    16:18                server

    Directory: C:\Users\...\astro-unocss-sample\dist\client

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
da---          12/01/2023    16:18                assets
-a---          03/01/2023    15:44            873 favicon.svg

    Directory: C:\Users\...\astro-unocss-sample\dist\client\assets

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          12/01/2023    16:18           2152 _@astrojs-ssr-virtual-entry.a3704025.css
-a---          12/01/2023    16:18           2057 index.c5bcaae5.css

    Directory: C:\Users\...\astro-unocss-sample\dist\server

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          12/01/2023    16:18         432584 entry.mjs
```

So, there is `entry.mjs` as usual, also there's `_@astrojs-ssr-virtual-entry.a3704025.css` as well, and looking to the name it is supposed to be like a virtual entry but I can't find any documentation how to deal with it.

Long story short:
- UnoCSS mode `dist-chunk` sounds very promising for MPA style applications.
- I can't make it work in Astro neither in `SSG` nor in `SSR + Node adapter` mode.
- Actually, I'm not sure if there is an issue in UnoCSS, or in Astro, or even in Vite, or in combination of any of them.

Any insights or help would be much appreciated.

**Comment by Yixuan-Wang** [comment]
Maybe you should try setting `injectEntry` to `false`, because the default behavior injects `uno.css` automatically into every .astro component. 
https://github.com/unocss/unocss/blob/654c7f973878c41cc1c07207f85e439d9fd5869f/packages/astro/src/index.ts#L20

However, if you are using Vue or Svelte SFCs, the behavior of `dist-chunk` will inject the import script into the outer scope instead of the `<script>` tag, which is still buggy.

**Comment by Yixuan-Wang** [comment]
This works for me (I am using `pnpm`):
```typescript
export default defineConfig({
  integrations: [
    Unocss({
      mode: "dist-chunk",
      injectEntry: process.env["NODE_ENV"] === "development",
    }),
  ],
});
```

Under dev mode, the Unocss Vite plugin uses `global` mode (even if `dist-chunk` is configured), which generates a `"uno.css"` file. The Astro integration automatically injects this import into every page. However, under production mode, there is no such `"uno.css"` file, as the generated style is directly emitted into each of the chunks, so this automatic injection will break the build. A workaround is to configure `injectEntry` manually according to the NODE_ENV mode.

PS: I'm not sure which tool injects this "NODE_ENV" env var (maybe `pnpm` or `astro-cli`), maybe you need some alternative logic to get this dev/prod mode.
PPS: I'm not sure whether this works for SSR.
